aten/src/ATen/native/cuda/jit_utils.cpp

```diff
@@ -808,6 +808,65 @@ std::string generate_code(
   return code;
 }
 
+// Creates directories recursively
+bool _r_mkdir(const std::string& dir) {
+  // Check if current dir exists
+  const char* p_dir = dir.c_str();
+  const bool dir_exists = (access(p_dir, F_OK) == 0);
+  if (dir_exists) {
+    return true;
+  }
+
+  // Try to create current directory
+#ifdef _WIN32
+  int ret = _mkdir(dir.c_str());
+#else
+  int ret = mkdir(dir.c_str(), S_IRWXU | S_IRWXG | S_IRWXO);
+#endif
+  // Success
+  if (ret == 0) {
+    return true;
+  }
+
+  // Find folder separator and check if we are at the top
+  int pos = dir.find_last_of("/\\");
+  if (pos == std::string::npos) {
+    return false;
+  }
+
+  // Try to create parent directory
+  if (!(_r_mkdir(dir.substr(0, pos)))) {
+    return false;
+  }
+
+  // Try to create complete path again
+#ifdef _WIN32
+  ret = _mkdir(dir.c_str());
+#else
+  ret = mkdir(dir.c_str(), S_IRWXU | S_IRWXG | S_IRWXO);
+#endif
+  return (ret==0);
+}
+
+// Creates direcotiries recusrively assuming that base exists
+bool r_mkdir_with_base(std::string& base, std::string& dir){
+  const char* p_base = base.c_str();
+  const bool base_exists = (access(p_base, F_OK) == 0);
+  if (!base_exists) {
+    return false;
+  }
+
+  // remove trailing '/' or '\\'
+  if ((base[base.size()-1]=='/') || base[base.size()-1]=='\\') {
+    base.pop_back();
+  }
+  if ((dir[dir.size()-1]=='/') || dir[dir.size()-1]=='\\') {
+    dir.pop_back();
+  }
+
+  return _r_mkdir(base+dir);
```

Comment:
Why remove the trailing slashes?

Doesn't `base` need to end with a slash if `dir` is appended to it like this?

Reply:
You are right that a forward slash is indeed needed and in the currently used code the `dir` provides it [here](https://github.com/pytorch/pytorch/blob/3a112ebb57b368ba83c3a658a761590d1924ccc7/aten/src/ATen/native/cuda/jit_utils.cpp#L835-L844).
However, it might be a better idea to either:
* allow two forward slashes in the path between `base` and `dir` (it should be ignored if it's not at the beginning of a path if I'm not mistaken) and remove the deletion in `base`
* check if `base` contains the `/` at the end or `dir` at the beginning and only keep one.

Let me know which approach sounds better.

Reply:
Aha, thank you. I suppose it's fine, then, and I don't know if it's interesting to try and "fix" user paths.

We should add a test that the cache is working, however. Not for this PR, necessarily, but @ngimel and I were thinking that a simple test which calls a jiterated kernel and then checks that the directory contains a file would be a good sanity check.